### PR TITLE
bird-watcher: add "for loop" concept

### PR DESCRIPTION
**Pull request by sachinmk27**
Fixes: #1617 

1. Updated config.json with concept and concept exercise
2. Added concept exercise directory with minimum valid spec
3. Added concept directory with about, introduction.

### CODE REVIEW

**Comment by junedev** [comment]
I marked this as a draft PR. Click on "ready for review" to when you are ready to change it back to a normal PR.

File: concepts/for-loops/about.md
Comment (jmrunkle) [inline concepts/for-loops/about.md:5]:
```suggestion
The `for` loop is one of the most commonly used statements to repeatedly execute some logic.
The basic `for` loop has three components separated by semicolons:
```

File: concepts/for-loops/about.md
Comment (jmrunkle) [inline concepts/for-loops/about.md:35]: super nitpicky: This is setup to look like a 1 to 10 iteration. We should probably make it actually do that.

```suggestion
for i := 1; i <= 10; i++ {
```

File: concepts/for-loops/about.md
Comment (jmrunkle) [inline concepts/for-loops/about.md:38]:
```suggestion
  fmt.Println(i)
```

Remove the unnecessary new-lines.

File: concepts/for-loops/about.md
Comment (jmrunkle) [inline concepts/for-loops/about.md:42]:
```suggestion
Defining the step is often done using the increment or decrement statement as shown in the example above.
```

This is also very nitpicky, but it turns out this is actually a statement because it does not return a value.

File: concepts/for-loops/about.md
Comment (jmrunkle) [inline concepts/for-loops/about.md:57]:
```suggestion
## The "while" loop
```

File: concepts/for-loops/about.md
Comment (jmrunkle) [inline concepts/for-loops/about.md:69]:
```suggestion
var sum int = 1
for sum < 1000 {
	sum += sum
}
fmt.Println(sum)
// Output: 1024
```

File: concepts/for-loops/about.md
Comment (jmrunkle) [inline concepts/for-loops/about.md:22]: Remove unneeded new-lines.

```suggestion
for initialization; condition; step {
  // code that is executed repeatedly as long as the condition is true
}
```

File: concepts/for-loops/about.md
Comment (jmrunkle) [inline concepts/for-loops/about.md:11]:
```suggestion
- the **init statement**: executed once before the first iteration
- the **condition expression**: evaluated before every iteration
- the **post statement**: executed at the end of every iteration
```

File: concepts/for-loops/about.md
Comment (jmrunkle) [inline concepts/for-loops/about.md:84]:
```suggestion
for n := 0; n <= 5; n++ {
  if n == 3 {
    break
  }
  fmt.Println(n)
}
// Output:
// 0
// 1
// 2
```

File: concepts/for-loops/about.md
Comment (jmrunkle) [inline concepts/for-loops/about.md:97]:
```suggestion
for n := 0; n <= 5; n++ {
  if n%2 == 0 {
    continue
  }
  fmt.Println(n)
}
// Output:
// 1
// 3
// 5
```

File: concepts/for-loops/about.md
Comment (jmrunkle) [inline concepts/for-loops/about.md:72]: I did not see labels mentioned. Were you planning to add that?

File: concepts/for-loops/links.json
Comment (jmrunkle) [inline concepts/for-loops/links.json:1]:
```suggestion
[]

```

Add a trailing new-line here.

File: config.json
Comment (jmrunkle) [inline config.json:219]:
```suggestion
          "for-loops"
```

The `increment-decrement` "concept" was actually included in `arithmetic-operations` so we should probably just delete it.

File: exercises/concept/bird-watcher/.docs/hints.md
Comment (jmrunkle) [inline exercises/concept/bird-watcher/.docs/hints.md:9]:
```suggestion
- Refer to the exercise introduction for an example how to use a for loop to iterate over an array.
- Use a helper variable to store the total count and increase that variable as you go through the array.
- Think about the correct initial value for that helper variable.
```

File: exercises/concept/bird-watcher/.docs/hints.md
Comment (jmrunkle) [inline exercises/concept/bird-watcher/.docs/hints.md:19]:
```suggestion
- This task is similar to the first one, you can copy your code as a starting point.
- Think about which indexes in the array you would need to take into account for week number 1 and 2, respectively.
- Now find a general way to calculate the first and the last index that should be considered.
- With that you can set up the for loop to only iterate over the relevant section of the array.
```

File: exercises/concept/bird-watcher/.docs/hints.md
Comment (jmrunkle) [inline exercises/concept/bird-watcher/.docs/hints.md:29]:
```suggestion
- Again you need to set up a for loop to iterate over the whole bird count array.
- This time you only need to visit every second entry in the array.
- Change the step so the counter variable is increased accordingly after each iteration.
- In the body of the for loop you can use the increment operator to change the value of an element in an array in place.
```

File: exercises/concept/bird-watcher/.docs/instructions.md
Comment (jmrunkle) [inline exercises/concept/bird-watcher/.docs/instructions.md:13]:
```suggestion
Let us start analyzing the data by getting a high level view.
Find out how many birds you counted in total since you started your logs.
```

File: exercises/concept/bird-watcher/.docs/instructions.md
Comment (jmrunkle) [inline exercises/concept/bird-watcher/.docs/instructions.md:15]:
```suggestion
Implement a function `TotalBirdCount` that accepts an array that contains the bird count per day.
It should return the total number of birds that you counted.
```

File: exercises/concept/bird-watcher/.docs/instructions.md
Comment (jmrunkle) [inline exercises/concept/bird-watcher/.docs/instructions.md:24]:
```suggestion
var birdsPerDay []int = []int{2, 5, 0, 7, 4, 1, 3, 0, 2, 5, 0, 1, 3, 1}
TotalBirdCount(birdsPerDay)
// => 34
```

File: exercises/concept/bird-watcher/.docs/instructions.md
Comment (jmrunkle) [inline exercises/concept/bird-watcher/.docs/instructions.md:33]:
```suggestion
It returns the total number of birds that you counted in that specific week.
You can assume weeks are always tracked completely.
```

File: exercises/concept/bird-watcher/.docs/instructions.md
Comment (jmrunkle) [inline exercises/concept/bird-watcher/.docs/instructions.md:42]:
```suggestion
var birdsPerDay []int = []int{2, 5, 0, 7, 4, 1, 3, 0, 2, 5, 0, 1, 3, 1}
BirdsInWeek(birdsPerDay, 2)
// => 12
```

File: exercises/concept/bird-watcher/.docs/instructions.md
Comment (jmrunkle) [inline exercises/concept/bird-watcher/.docs/instructions.md:64]:
```suggestion
var birdsPerDay []int = []int{2, 5, 0, 7, 4, 1}
FixBirdCountLog(birdsPerDay)
// => [3 5 1 7 5 1]
```

File: exercises/concept/bird-watcher/.meta/config.json
Comment (jmrunkle) [inline exercises/concept/bird-watcher/.meta/config.json:2]:
```suggestion
  "blurb": "Professionalize counting the birds in your garden with for loops.",
```

**Review by jmrunkle — DISMISSED** [review]
Thanks! Many of the documentation comments also need to be applied to the concept and exercise introduction.md files, but I only commented on about.md.

File: concepts/for-loops/about.md
Comment (ErikSchierboom) [inline concepts/for-loops/about.md:108]: Maybe simplify this to:

```suggestion
  // Endless loop...
```

File: concepts/for-loops/introduction.md
Comment (ErikSchierboom) [inline concepts/for-loops/introduction.md:9]: Applying what was suggested on the `about.md` document above:

```suggestion
- the **init statement**: executed before the first iteration
- the **condition expression**: evaluated before every iteration
- the **post statement**: executed at the end of every iteration
```

File: concepts/for-loops/introduction.md
Comment (ErikSchierboom) [inline concepts/for-loops/introduction.md:3]: Well put!

File: concepts/for-loops/introduction.md
Comment (ErikSchierboom) [inline concepts/for-loops/introduction.md:20]:
```suggestion
for initialization; condition; step {
  // code that is executed repeatedly as long as the condition is true
}
```

File: concepts/for-loops/introduction.md
Comment (ErikSchierboom) [inline concepts/for-loops/introduction.md:38]:
```suggestion
for i := 1; i < 10; i++ {
  fmt.Println(i)
}
```

File: concepts/for-loops/introduction.md
Comment (ErikSchierboom) [inline concepts/for-loops/introduction.md:41]:
```suggestion
Defining the step is often done using an increment or decrement statement, as shown in the example above.
```

File: exercises/concept/bird-watcher/.docs/instructions.md
Comment (ErikSchierboom) [inline exercises/concept/bird-watcher/.docs/instructions.md:8]:
```suggestion
Usually you use a tally in a notebook to count the birds, but to better work with your data, you've digitalized the bird counts for the past weeks.
```

File: exercises/concept/bird-watcher/.docs/instructions.md
Comment (ErikSchierboom) [inline exercises/concept/bird-watcher/.docs/instructions.md:9]: I'd remove this in favor of the descriptions per task.

```suggestion
```

File: exercises/concept/bird-watcher/bird_watcher.go
Comment (ErikSchierboom) [inline exercises/concept/bird-watcher/bird_watcher.go:20]:
```suggestion
```

File: exercises/concept/bird-watcher/bird_watcher.go
Comment (ErikSchierboom) [inline exercises/concept/bird-watcher/bird_watcher.go:13]:
```suggestion
```

File: exercises/concept/bird-watcher/bird_watcher.go
Comment (ErikSchierboom) [inline exercises/concept/bird-watcher/bird_watcher.go:4]:
```suggestion
// TotalBirdCount return the total bird count by summing
// the individual day's counts
```

File: exercises/concept/bird-watcher/bird_watcher.go
Comment (ErikSchierboom) [inline exercises/concept/bird-watcher/bird_watcher.go:17]: Maybe leave out the implementation detail about arrays?

```suggestion
// FixBirdCountLog returns the bird counts after correcting
// the bird counts for alternate days
```

File: exercises/concept/bird-watcher/.docs/introduction.md
Comment (ErikSchierboom) [inline exercises/concept/bird-watcher/.docs/introduction.md:11]:
```suggestion
- the **init statement**: executed before the first iteration
- the **condition expression**: evaluated before every iteration
- the **post statement**: executed at the end of every iteration
```

File: exercises/concept/bird-watcher/.docs/introduction.md
Comment (ErikSchierboom) [inline exercises/concept/bird-watcher/.docs/introduction.md:22]:
```suggestion
for initialization; condition; step {
  // code that is executed repeatedly as long as the condition is true
}
```

File: exercises/concept/bird-watcher/.docs/introduction.md
Comment (ErikSchierboom) [inline exercises/concept/bird-watcher/.docs/introduction.md:42]:
```suggestion
var list []int = []int{1,2,3}

for i := 1; i < len(list); i++ {
  fmt.Println(i)
}
```

File: exercises/concept/bird-watcher/.docs/introduction.md
Comment (ErikSchierboom) [inline exercises/concept/bird-watcher/.docs/introduction.md:45]:
```suggestion
Defining the step is often done using an increment or decrement statement as shown in the example above.
```

**Review by ErikSchierboom — COMMENTED** [review]
Thanks for working on this

**Comment by jmrunkle** [comment]
@sachinmk27 - any chance you will have time to work on this? This is also blocking ErikSchierboom's new exercise interest-is-interesting (unless we remove the prerequisite and add it later).

**Comment by sachinmk27** [comment]
@jmrunkle Unfortunately I will not be able to make time for the next 4-5 days. Kindly reassign it since it's already delayed.

**Comment by junedev** [comment]
@sachinmk27 Could you give me an update again whether you want to finish this PR or whether we should re-assign the task?

**Comment by sachinmk27** [comment]
Pls reassign the task. Sorry, I couldn't make time for it.

**Comment by andrerfcsantos** [comment]
> Pls reassign the task. Sorry, I couldn't make time for it.

No worries, thanks for letting us know :)

File: exercises/concept/bird-watcher/bird_watcher_test.go
Comment (junedev) [inline exercises/concept/bird-watcher/bird_watcher_test.go:97]: this should not be part of the test data table, you can set up anything you need in the separate subtest that you have below

File: exercises/concept/bird-watcher/bird_watcher_test.go
Comment (junedev) [inline exercises/concept/bird-watcher/bird_watcher_test.go:113]: I am not sure the test runner can handle this extra subtest, I will check later

**Review by junedev — COMMENTED** [review]
Left some ideas for improvement.

**Comment by junedev** [comment]
@andrerfcsantos I only reviewed about.md so fixes you do there need to be applied to the other files as well. ... you probably know the drill. :wink:

File: exercises/concept/bird-watcher/.docs/instructions.md
Comment (andrerfcsantos) [reply exercises/concept/bird-watcher/.docs/instructions.md]: Fixed. I noticed the hints also had some references to arrays, I also replaced them.

File: concepts/for-loops/about.md
Comment (andrerfcsantos) [reply concepts/for-loops/about.md]: Just reworded it so the whole word is avoided.

File: concepts/for-loops/about.md
Comment (andrerfcsantos) [reply concepts/for-loops/about.md]: Removed the "will never" end bit.

**Comment by andrerfcsantos** [comment]
@junedev Gotcha. Ready to review again.

File: exercises/concept/bird-watcher/bird_watcher_test.go
Comment (junedev) [reply exercises/concept/bird-watcher/bird_watcher_test.go:108]: not sure why this was resolved, I checked it and indeed, the test code extraction cannot handle this separate subtest, the user will see the full code of the whole TestFixBirdCount test which is very confusing

Since I think fixing the test runner would take quite some time, I would recommend to move this subtest into a separate (non-sub) test TestFixBirdCount_DoesNotCreateNewSlice.

Comment (andrerfcsantos) [reply exercises/concept/bird-watcher/bird_watcher_test.go:109]: Sorry, this was me marking it as resolved. I thought you were going to check after the merge.

Anyway, I made the change now.

Comment (junedev) [reply exercises/concept/bird-watcher/bird_watcher_test.go:109]: Ah ok, no, I don't like to merge it without proper test runner code extraction because then usually we get confused students opening issues.

I looked at the latest version you pushed. The subtest inside `TestFixBirdCountDoesNotCreateNewSlice` is obsolete now, there is only this one and you already have the name in the name of the parent test.

File: exercises/concept/bird-watcher/.docs/instructions.md
Comment (junedev) [reply exercises/concept/bird-watcher/.docs/instructions.md]: you only fixed this in one place, should be applied consistently for all examples

File: exercises/concept/bird-watcher/bird_watcher_test.go
Comment (andrerfcsantos) [reply exercises/concept/bird-watcher/bird_watcher_test.go]: Just removed the subtest.

**Review by junedev — APPROVED** [review]

**Comment by junedev** [comment]
@andrerfcsantos The only thing left is this comment: https://github.com/exercism/go/pull/1775#discussion_r739879670
I approved so you can merge yourself after that last fix.

File: exercises/concept/bird-watcher/.docs/instructions.md
Comment (andrerfcsantos) [reply exercises/concept/bird-watcher/.docs/instructions.md]: Fixed now.

**Pull request merged**
